src/plugins/swap.c

```diff
@@ -198,56 +201,150 @@ gboolean bd_swap_swapon (const gchar *device, gint priority, GError **error) {
 
     bd_utils_report_progress (progress_id, 0, "Analysing the swap device");
     /* check the device if it is an activatable swap */
-    dev_file = g_io_channel_new_file (device, "r", error);
-    if (!dev_file) {
-        /* error is already populated */
+    probe = blkid_new_probe ();
+    if (!probe) {
+        g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_UNKNOWN_STATE,
+                     "Failed to create a new probe");
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return FALSE;
+    }
+
+    fd = open (device, O_RDONLY|O_CLOEXEC);
+    if (fd == -1) {
+        g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_UNKNOWN_STATE,
+                     "Failed to open the device '%s'", device);
         bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
         return FALSE;
     }
 
-    page_size = getpagesize ();
-    page_size = MAX (2048, page_size);
-    io_status = g_io_channel_seek_position (dev_file, page_size - 10, G_SEEK_SET, &tmp_error);
-    if (io_status != G_IO_STATUS_NORMAL) {
+    /* we may need to try mutliple times with some delays in case the SWAPDEVICE is
+       busy at the very moment */
+    for (n_try=5, status=-1; (status != 0) && (n_try > 0); n_try--) {
+        status = blkid_probe_set_device (probe, fd, 0, 0);
+        if (status != 0)
+            g_usleep (100 * 1000); /* microseconds */
+    }
+    if (status != 0) {
         g_set_error (error, BD_SWAP_ERROR, BD_SWAP_ERROR_UNKNOWN_STATE,
-                     "Failed to determine device's state: %s", tmp_error->message);
-        g_clear_error (&tmp_error);
-        g_io_channel_shutdown (dev_file, FALSE, &tmp_error);
+                     "Failed to create a probe for the device '%s'", device);
         bd_utils_report_finished (progress_id, (*error)->message);
+        blkid_free_probe (probe);
+        close (fd);
         return FALSE;
     }
 
-    io_status = g_io_channel_read_chars (dev_file, dev_status, 10, &num_read, &tmp_error);
-    if ((io_status != G_IO_STATUS_NORMAL) || (num_read != 10)) {
+    blkid_probe_enable_partitions (probe, 1);
+    blkid_probe_set_partitions_flags (probe, BLKID_PARTS_MAGIC);
+    blkid_probe_enable_superblocks (probe, 1);
+    blkid_probe_set_superblocks_flags (probe, BLKID_SUBLKS_USAGE | BLKID_SUBLKS_TYPE |
+                                              BLKID_SUBLKS_MAGIC | BLKID_SUBLKS_BADCSUM);
```

Comment:
Do we really want these features? Not sure what *superblocks* here are, but we are definitely not probing for partitions.